test/difference.jl

```diff
@@ -0,0 +1,57 @@
+function test_difference(ε::Real, x, dx)
+    y = x + ε * dx
+    dx_diff = difference(ε, y, x)
+    @test typeof(dx) == typeof(dx_diff)
```

Comment:
This testing is a bit weak. Ideally we would test that the definition in the docstring holds. `==` is out of the question, and most types don't have `isapprox` defined, so it's hard to know what to do. Suggestions appreciated.

Reply:
what types here cause a problem for `isapprox`?

presumably we should be okay to define `isapprox(::Zero, ::Zero) = true` if that's all we need?

Reply:
Well it's certainly an issue for all composites, but purely in the sense that we've not defined it. We could define it. Maybe that's a good idea?

Reply:
I've opened https://github.com/JuliaDiff/ChainRulesCore.jl/issues/184 about this.

Reply:
Thanks. Can you also add a TODO to the code?
```suggestion
    # TODO: `@test isapprox(dx, dx_diff)` once `isapprox` is defined appropriately
    # see https://github.com/JuliaDiff/ChainRulesCore.jl/issues/184
    @test typeof(dx) == typeof(dx_diff)
```